### PR TITLE
Add regex matching to no-commit-to-branch hook

### PR DESCRIPTION
**Pull request by marcjay**
Add wildcard matching to no-commit-to-branch hook, so that commits can be blocked on, for example, all release branches with 'release/*'

Addresses: #375

### CODE REVIEW

File: pre_commit_hooks/no_commit_to_branch.py [hunk 1]
```diff
@@ -10,14 +10,17 @@
 from pre_commit_hooks.util import cmd_output
 
 
-def is_on_branch(protected):  # type: (Set[str]) -> bool
+def is_on_branch(protected, patterns=set()):
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: the default value here should be immutable (such as `frozenset()`)

Comment (marcjay) [reply pre_commit_hooks/no_commit_to_branch.py]: Cheers for the review :smile:! Have made the change. Also had to change the calling function to satisfy the type checker

**Comment by asottile** [comment]
cool! I switched it to `AbstractSet` to appease mypy

**Comment by marcjay** [comment]
Thanks! Appreciate the crash-course in mypy and cheers for fixing

**Comment by asottile** [comment]
And thank you for the pull request 🎉

**Pull request merged**